From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malaya Kumar Rout <malayarout91@gmail.com>
Date: Sat, 4 Jul 2026 15:04:28 +0530
Subject: [PATCH] time/namespace: Validate nanosecond field in
 proc_timens_set_offset()

[ Upstream commit 06aba58e58492d2b8eae059274caed29025ea96e ]

The function validates tv_sec to be within [-KTIME_SEC_MAX, KTIME_SEC_MAX]
but never validates that tv_nsec is within the valid range of
[0, NSEC_PER_SEC-1] before using it in timespec64_add().

timespec64_add() expects both timespec64 structures to have normalized
values with tv_nsec in the range [0, 999999999]. If off->val.tv_nsec
contains invalid values (negative or >= NSEC_PER_SEC), it could lead to
incorrect calculations or unexpected behavior.

Add validation to ensure tv_nsec is within the valid range before
performing the addition.

Fixes: 04a8682a71be ("fs/proc: Introduce /proc/pid/timens_offsets")
Signed-off-by: Malaya Kumar Rout <malayarout91@gmail.com>
Signed-off-by: Thomas Gleixner <tglx@kernel.org>
Link: https://patch.msgid.link/20260704093429.89350-1-malayarout91@gmail.com
Signed-off-by: Sasha Levin <sashal@kernel.org>
(cherry picked from commit 1f0c3c0c36c7aedf55a60d9c71c4bd258b8937ce)
CVE-2026-93146
[TL: backport: proc_timens_set_offset() still uses the err and goto out
 error handling here, adapt the new check to it]
Signed-off-by: Thomas Lamprecht <t.lamprecht@proxmox.com>
---
 kernel/time/namespace.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/kernel/time/namespace.c b/kernel/time/namespace.c
index 652744e00eb4c0b1bf29f0098f17543b93b56e9d..51f51abef23b5f18a8ccacd09b6999ce6c20ddc8 100644
--- a/kernel/time/namespace.c
+++ b/kernel/time/namespace.c
@@ -419,6 +419,11 @@ int proc_timens_set_offset(struct file *file, struct task_struct *p,
 		    off->val.tv_sec < -KTIME_SEC_MAX)
 			goto out;
 
+		if (off->val.tv_nsec < 0 || off->val.tv_nsec >= NSEC_PER_SEC) {
+			err = -EINVAL;
+			goto out;
+		}
+
 		tp = timespec64_add(tp, off->val);
 		/*
 		 * KTIME_SEC_MAX is divided by 2 to be sure that KTIME_MAX is
